code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/TokenUtils.java

```diff
@@ -0,0 +1,187 @@
+        } catch (TokenException e) { // in case of invalid token
+            return null;
+        }
+        if (jsonData == null) {
+            throw new ClientVisibleException(ResponseCodes.BAD_REQUEST, ACCESS_TOKEN_INVALID,
+                    "Json Web Token invalid.", null);
+        }
+        String tokenTypeActual = (String) jsonData.get(TOKEN);
+        if (!StringUtils.equals(tokenType, tokenTypeActual)) {
+            return null;
+        }
+        if (!isAllowed(jsonData)) {
+            throw new ClientVisibleException(ResponseCodes.UNAUTHORIZED);
+        }
+        return jsonData;
+    }
+
+    public String getJWT() {
+        ApiRequest request = ApiContext.getContext().getApiRequest();
+        String jwt = (String) request.getAttribute(tokenType());
+        if (StringUtils.isNotBlank(jwt) && getJsonData(jwt, tokenType()) == null) {
+            throw new ClientVisibleException(ResponseCodes.BAD_REQUEST, ValidationErrorCodes.INVALID_FORMAT,
+                    "Token malformed after retrieval.", null);
+        }
+        return jwt;
+    }
+
+    protected abstract boolean isAllowed(Map<String, Object> jsonData);
+
+    public Set<Identity> identities() {
+        Map<String, Object> jsonData = getJsonData();
+        if (jsonData == null) {
+            return new HashSet<>();
+        }
+        return identities(jsonData);
+    }
+
+    protected abstract Set<Identity> identities(Map<String, Object> jsonData);
+
+    public Set<Identity> getIdentities() {
+        return identities();
+    }
+
+    public boolean findAndSetJWT() {
+        ApiRequest request = ApiContext.getContext().getApiRequest();
+        String jwt = (String) request.getAttribute(tokenType());
+        if (StringUtils.isNotBlank(jwt) && getJsonData(jwt, tokenType()) != null) {
+            return true;
+        }
+        if (StringUtils.isBlank(jwt)) {
+            Cookie[] cookies = request.getServletContext().getRequest().getCookies();
+            if (cookies != null) {
+                for (Cookie cookie : cookies) {
+                    if (cookie.getName().equalsIgnoreCase(TOKEN)
+                            && StringUtils.isNotBlank(cookie.getValue())) {
+                        jwt = cookie.getValue();
+                        break;
+                    }
+                }
+            }
+        }
+        if (StringUtils.isBlank(jwt)) {
+            jwt = request.getServletContext().getRequest().getHeader(ProjectConstants.AUTH_HEADER);
+        }
+        if (StringUtils.isBlank(jwt)) {
+            jwt = request.getServletContext().getRequest().getParameter(TOKEN);
+        }
+        if (getJsonData(jwt, tokenType()) != null) {
+            request.setAttribute(tokenType(), jwt);
+            return true;
+        }
+        return false;
+    }
+
+    public boolean isAllowed(List<String> idList, Set<Identity> identities) {
+        boolean hasAccessToAProject = authDao.hasAccessToAnyProject(identities, false, null);
```

Comment:
Save a DAO hit when in unrestricted mode by putting the hasAccessToAnyProject under the restricted case.